File: core/std/java/src/fan/std/Lock.java
package fan.std;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;

import fan.sys.FanObj;
import fan.sys.Func;
import fanx.main.Sys;
import fanx.main.Type;

public class Lock extends FanObj {
	java.util.concurrent.locks.Lock lock;
	
	public static Lock make() {
		Lock l = new Lock();
		l.lock = new ReentrantLock();
		return l;
	}
	
	private static Type type = null;
	public Type typeof() { if (type == null) { type = Sys.findType("std::Lock"); } return type;  }
	
	public boolean tryLock(long nanoTime) {
		try {
			return lock.tryLock(nanoTime, TimeUnit.NANOSECONDS);
		} catch (InterruptedException e) {
			return false;
		}
	}
	
	public void lock() {
		lock.lock();
	}
	
	public void unlock() {
		lock.unlock();
	}
	
	public Object sync(Func c) {
		try {
			lock.lock();
			return c.call();
		}
		finally {
			lock.unlock();
		}
	}
}

File: library/std/js/Lock.js



fan.std.Lock = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.Lock.prototype.$ctor = function() {}
fan.std.Lock.prototype.$typeof = function() { return fan.std.Lock.$type; }

fan.std.Lock.make = function() {
	var self = new fan.std.Lock();
	self.m_lock = false;
	return self;
}

fan.std.Lock.prototype.tryLock = function(nanoTime) {
	if (this.m_lock) {
		return false;
	}
	return true;
}

fan.std.Lock.prototype.lock = function() {
	this.m_lock = true;
}

fan.std.Lock.prototype.unlock = function() {
	this.m_lock = false;
}

fan.std.Lock.prototype.sync = function(c) {
	try {
		this.lock();
		return c.call();
	}
	finally {
		this.unlock();
	}
}
